packages/server/src/api/com/atproto/repo.ts

```diff
@@ -117,30 +118,49 @@ export default function (server: Server) {
         }
       }
     }
-    const repo = await locals.loadRepo(res, did)
-    if (!repo) {
-      throw new InvalidRequestError(
-        `${did} is not a registered repo on this server`,
-      )
-    }
-    const prevCid = repo.cid
-    // generate TIDs for create actions without an rkey
-    const test = tx.writes.map((write) => ({
-      ...write,
-      rkey: write.rkey || TID.nextStr(),
-    }))
-    await repo.batchWrite(test)
-    // @TODO: do something better here instead of rescanning for diff
-    const diff = await repo.verifyUpdates(prevCid, repo.cid)
-    try {
-      await repoDiff.processDiff(db, repo, diff)
-    } catch (err) {
-      logger.warn({ did, err }, 'failed to index batch write')
-      if (validate) {
-        throw new InvalidRequestError(`Could not index record: ${err}`)
+    await db.transaction(async (dbTxn) => {
+      const currRoot = await dbTxn.getRepoRoot(did, true)
+      if (!currRoot) {
+        throw new InvalidRequestError(
+          `${did} is not a registered repo on this server`,
+        )
       }
-    }
-    await db.updateRepoRoot(did, repo.cid)
+      const cidWriteOps: CidWriteOp[] = await Promise.all(
+        tx.writes.map(async (write) => {
+          if (write.action === 'create') {
+            const cid = await blockstore.put(write.value)
+            const rkey = write.rkey || TID.nextStr()
+            const uri = new AtUri(`${did}/${write.collection}/${rkey}`)
+            await dbTxn.indexRecord(uri, cid, write.value)
+            return {
+              action: 'create',
+              collection: write.collection,
+              rkey,
+              cid,
+            }
+          } else if (write.action === 'delete') {
+            const uri = new AtUri(`${did}/${write.collection}/${write.rkey}`)
+            await dbTxn.deleteRecord(uri)
+            return write
+          } else {
+            throw new InvalidRequestError(
+              `Action not supported: ${write.action}`,
+            )
+          }
+        }),
+      )
+
+      const repo = await RepoStructure.load(blockstore, currRoot)
+      await repo
+        .stageUpdate(cidWriteOps)
+        .createCommit(authStore, async (prev, curr) => {
+          const success = await db.updateRepoRoot(did, curr, prev)
+          if (!success) {
+            throw new Error('Could not update repo root')
```

Comment:
We might consider turning this into a 409 Conflict for HTTP-land.

Reply:
Still this feels like a special enough case (& encountered in several spots) that we may want a special error type to handle it?

Reply:
That makes a lot of sense.  Barring a special error type, maybe it's the kind of thing we'd want to log.

Reply:
yeah that's a good idea